### PR TITLE
ENH: Use new ``DerivativesDataSink`` from NiWorkflows 1.2.0

### PR DESCRIPTION
**Pull request by oesteban**
This PR integrates the new API that will be released with 1.2.0.

References: nipreps/niworkflows#507

### CODE REVIEW

File: fmriprep/workflows/bold/confounds.py [hunk 1]
```diff
@@ -509,7 +509,7 @@ def init_carpetplot_wf(mem_gb, metadata, cifti_output, name="bold_carpet_wf"):
             ('framewise_displacement', 'mm', 'FD')]),
         name='conf_plot', mem_gb=mem_gb)
     ds_report_bold_conf = pe.Node(
-        DerivativesDataSink(desc='carpetplot', keep_dtype=True),
+        DerivativesDataSink(desc='carpetplot', datatype="figures", extension="svg"),
```
Comment (oesteban) [inline fmriprep/workflows/bold/confounds.py:512]: For some reason, the name of the intermediate carpetplot generated for surfaces has the form `.carpetplot.svg`, and that "carpetplot" is interpreted as extension.

File: fmriprep/workflows/bold/outputs.py [hunk 2]
```diff
@@ -256,60 +256,133 @@ def init_func_derivatives_wf(
 
     fs_outputs = spaces.cached.get_fs_spaces()
     if freesurfer and fs_outputs:
-        from niworkflows.interfaces.surf import GiftiNameSource
```
Comment (oesteban) [inline fmriprep/workflows/bold/outputs.py:259]: I haven't removed this for backwards compatibility, but the new Path2BIDS should simplify its logic (no need anymore for building up the new pattern).

File: fmriprep/workflows/bold/outputs.py [hunk 3]
```diff
@@ -256,60 +256,133 @@ def init_func_derivatives_wf(
...
+                                        ('source_file', 'source_file'),
+                                        (('cifti_metadata', _get_surface), 'space'),
+                                        ('cifti_density', 'density'),
+                                        (('cifti_metadata', _read_json), 'meta_dict')])
         ])
 
     return workflow
+
+
+def init_bold_preproc_report_wf(mem_gb, reportlets_dir, name='bold_preproc_report_wf'):
```
Comment (oesteban) [inline fmriprep/workflows/bold/outputs.py:304]: This change is a bit out of scope. This workflow is not used anywhere (I wrote it when there were plans to show a before/after reportlet without/with preprocessing. This will potentially be used at some point, and I thought it was more findable here in the outputs module.

File: setup.cfg [hunk 4]
```diff
@@ -12,6 +12,7 @@ description = fMRIPrep is a robust and easy-to-use pipeline for preprocessing of
 license = 3-clause BSD
 long_description = file:long_description.rst
 long_description_content_type = text/x-rst; charset=UTF-8
+name= fmriprep
```
Comment (oesteban) [inline setup.cfg:15]: this, the `setup_requires`, and the changes to `setup.py` are opportunistic, at some point, I needed to trigger a build and it seemed a nice cleanup pretty harmless.

File: setup.cfg [hunk 5]
```diff
@@ -27,21 +28,23 @@ install_requires =
...
+    smriprep @ git+https://github.com/oesteban/smriprep.git@enh/new-datasinks
     tedana >= 0.0.5
     templateflow >= 0.5.2
     toml
 test_requires =
     coverage
     codecov
     pytest
+setup_requires =
+    setuptools >= 40.8.0
```
Comment (oesteban) [inline setup.cfg:47]: I've kept 40.8.0 because it was already the pinned version.

While struggling with some sdist building issues (I finally discovered that ``setuptools-scm``, now a dependency to sdcflows, basically defenestrates the ``MANIFEST.in`` and auto-includes anything tracked by git), incidentally I learned the key versions of setuptools to work with `setup.cfg`:

```
setup_requires =
    # setup.cfg supported by +30.3.0
    setuptools >= 38.3.0  # +38.3 contains most of bugfixes
```
Noted here: https://github.com/nipreps/niworkflows/pull/507/files#diff-380c6a8ebbbce17d55d50ef17d3cf906R45-R47

This is not something I really care much (as long as we make sure `setuptools >= 38.3.0`).

File: setup.py [hunk 6]
```diff
@@ -1,23 +1,14 @@
...
-        cmdclass=versioneer.get_cmdclass(),
-        setup_requires=SETUP_REQUIRES,
-    )
+    setupargs = {
+        "version": versioneer.get_version(),
+        "cmdclass": versioneer.get_cmdclass(),
+    }
+    if "bdist_wheel" in sys.argv:
+        setupargs["setup_requires"] = ["setuptools >= 40.8", "wheel"]
+    setup(**setupargs)
```
Comment (oesteban) [inline setup.py:14]: this felt cleaner, happy to back up if it is annoying.

File: fmriprep/workflows/bold/base.py [hunk 7]
```diff
@@ -826,7 +826,7 @@ def init_func_preproc_wf(bold_file):
     # Fill-in datasinks of reportlets seen so far
     for node in workflow.list_node_names():
         if node.split('.')[-1].startswith('ds_report'):
-            workflow.get_node(node).inputs.base_directory = reportlets_dir
+            workflow.get_node(node).inputs.base_directory = output_dir
```
[the remark this answers is not in the document]
Comment (effigies) [reply fmriprep/workflows/bold/base.py:829]: This will also put HTML snippets in `figures/`, right?

File: fmriprep/workflows/bold/outputs.py [hunk 8]
```diff
@@ -240,76 +247,154 @@ def init_func_derivatives_wf(
...
-                                      ('source_file', 'source_file')]),
-            (name_cifti, cifti_bolds, [('out_name', 'suffix')]),
-            (name_cifti, cifti_key, [('out_name', 'suffix')]),
-            (inputnode, cifti_key, [('source_file', 'source_file'),
-                                    ('cifti_metadata', 'in_file')]),
+            (inputnode, ds_bold_cifti, [(('bold_cifti', _unlist), 'in_file'),
+                                        ('source_file', 'source_file'),
+                                        (('cifti_metadata', _get_surface), 'space'),
+                                        ('cifti_density', 'density'),
+                                        (('cifti_metadata', _read_json), 'meta_dict')])
```
Comment (effigies) [inline fmriprep/workflows/bold/outputs.py:310]: We've gone from `MapNode`s to one node. Is this safe, if somebody asks for two densities?

File: fmriprep/workflows/bold/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (oesteban) [reply fmriprep/workflows/bold/base.py:829]: Correct

File: fmriprep/workflows/bold/outputs.py (hunk 8 above)
Comment (oesteban) [reply fmriprep/workflows/bold/outputs.py:310]: @mgxd correct me if I'm wrong, but ``--cifti-outputs`` only accepts one density, right? And more certainly, the sampling workflow `init_bold_grayords_wf` definitely accepts only one density value.

So this should be safe because the just can't ask for two densities.

File: .circleci/ds005_fasttrack_outputs.txt [hunk 9]
```diff
@@ -9,69 +9,69 @@ fmriprep/logs/CITATION.md
...
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-aseg_dseg.nii.gz
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-brain_mask.json
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-brain_mask.nii.gz
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-preproc_bold.json
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-preproc_bold.nii.gz
+fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-1_AROMAnoiseICs.csv
+fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-1_desc-confounds_regressors.json
+fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-1_desc-confounds_regressors.tsv
+fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-1_desc-MELODIC_mixing.tsv
+fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-1_space-fsaverage5_hemi-L_bold.func.gii
```
[the remark this answers is not in the document]
Comment (oesteban) [reply .circleci/ds005_fasttrack_outputs.txt:16]: @mgxd @effigies, just double-checking you did not oversee this one, as it could be controversial.

File: fmriprep/workflows/bold/outputs.py (hunk 8 above)
Comment (mgxd) [reply fmriprep/workflows/bold/outputs.py:310]: yep, currently only one density is allowed per fmriprep call

File: fmriprep/workflows/bold/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (effigies) [reply fmriprep/workflows/bold/base.py:829]: Do we want that?

[the remark this answers is not in the document]
Comment (oesteban) [reply fmriprep/workflows/bold/base.py:829]: Well, what's the harm?

On the pro side - when I wrote the sMRIPrep fast-track I found an issue in generating the T1w summary (because the actual inputs were not available anymore)(*). So I had to set up an alternative that says: "the T1w was reused from a previous run".

Now, we could remove that alternative, as you indeed have the HTML excerpt to integrate in the report.

(*) Actually, this stands for any of the reportlets - there's a fair amount of logic in smriprep to regenerate all the reportlets if you are on the fast track.

File: .circleci/ds005_fasttrack_outputs.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (effigies) [reply .circleci/ds005_fasttrack_outputs.txt:16]: Hmm. Yeah, maybe let's not do this until we finalize derivatives, to minimize the number of times we change derivatives.

[the remark this answers is not in the document]
Comment (oesteban) [reply .circleci/ds005_fasttrack_outputs.txt:16]: The default pattern is already written in niworkflows (which means, rolling back requires one more PR cycle). In addition, the new pattern would be more regular so hopefully more likely to be in line with the finalized derivatives.

If the reason to roll back is just practical (i.e., the reason to introduce this change is not counterargued), I'd rather keep it.

[the remark this answers is not in the document]
Comment (effigies) [reply .circleci/ds005_fasttrack_outputs.txt:16]: I'm okay with this. It looks like it's not covered explicitly anywhere in the derivatives draft, so we should make a point to argue for it.

File: .circleci/ds005_fasttrack_outputs.txt [hunk 10]
```diff
@@ -9,69 +9,69 @@ fmriprep/logs/CITATION.md
...
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-MNI152NLin6Asym_desc-preproc_bold.nii.gz
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-MNI152NLin6Asym_desc-smoothAROMAnonaggr_bold.nii.gz
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_boldref.nii.gz
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-aparcaseg_dseg.nii.gz
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-aseg_dseg.nii.gz
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-brain_mask.json
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-brain_mask.nii.gz
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-preproc_bold.json
-fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-02_space-T1w_desc-preproc_bold.nii.gz
+fmriprep/sub-01/func/sub-01_task-mixedgamblestask_run-1_AROMAnoiseICs.csv
```
[the remark this answers is not in the document]
Comment (effigies) [reply .circleci/ds005_fasttrack_outputs.txt:12]: Let's do a separate PR, because we need to talk through the consequences with interested users.

File: fmriprep/workflows/bold/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (effigies) [reply fmriprep/workflows/bold/base.py:829]: Okay.

**Comment by oesteban** [comment]
This is looking good. Can we just merge now? Last call for last minute reviews?

**Comment by effigies** [comment]
Let's do it.

**Comment by oesteban** [comment]
Okay - 
brace yourselves

(sorry, couldn't hold it)
